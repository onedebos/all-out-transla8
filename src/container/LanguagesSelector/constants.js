export default {
	srcLangOptions: [{ langCode: 'EN', lang: 'English' }],

	targetLangOptions: [
		{ langCode: 'BG', lang: 'Bulgarian' },
		{ langCode: 'CS', lang: 'Czech' },
		{ langCode: 'DA', lang: 'Danish' },
		{ langCode: 'DE', lang: 'German' },
		{ langCode: 'EL', lang: 'Greek' },
		{ langCode: 'EN-GB', lang: 'English (British)' },
		{ langCode: 'EN-US', lang: 'English (American)' },
		// "EN" - English (unspecified variant for backward compatibility; please select EN-GB or EN-US instead)
		{ langCode: 'ES', lang: 'Spanish' },
		{ langCode: 'ET', lang: 'Estonian' },
		{ langCode: 'FI', lang: 'Finnish' },
		{ langCode: 'FR', lang: 'French' },
		{ langCode: 'HU', lang: 'Hungarian' },
		{ langCode: 'IT', lang: 'Italian' },
		{ langCode: 'JA', lang: 'Japanese' },
		{ langCode: 'LT', lang: 'Lithuanian' },
		{ langCode: 'LV', lang: 'Latvian' },
		{ langCode: 'NL', lang: 'Dutch' },
		{ langCode: 'PL', lang: 'Polish' },
		{ langCode: 'PT-PT', lang: 'Portuguese' },
		{ langCode: 'PT-BR', lang: 'Portuguese (Brazilian)' },
		// {"PT" : Portuguese (unspecified variant for backward compatibility; please select PT:PT or PT:BR instead)}
		{ langCode: 'RO', lang: 'Romanian' },
		{ langCode: 'RU', lang: 'Russian' },
		{ langCode: 'SK', lang: 'Slovak' },
		{ langCode: 'SL', lang: 'Slovenian' },
		{ langCode: 'SV', lang: 'Swedish' },
		{ langCode: 'ZH', lang: 'Chinese' },
	],

	SRC_LANG: 'Source Language',
	TRGT_LANG: 'Target Language',
};
