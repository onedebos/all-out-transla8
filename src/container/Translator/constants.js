export default {
    NO_SRC_TEXT_ERR: "Enter a word to translate...",
    NO_TRGT_LANG_ERR: "Select a language to translate into."
}