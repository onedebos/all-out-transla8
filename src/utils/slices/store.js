import { configureStore } from "@reduxjs/toolkit";

import translatorReducer from "./translator/translatorSlice";

export default configureStore({
  reducer: {
    translator: translatorReducer,
  },
  devTools: process.env.NODE_ENV !== "production" ? true : false,
});