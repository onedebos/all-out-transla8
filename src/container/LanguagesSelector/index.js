import LanguageSelect from '../../components/LanguageSelect';
import constants from './constants';
import styled from 'styled-components';
import { useTranslatorSlice } from '../../utils/slices/translator/useTranslatorSlice';
import { useEffect } from 'react';

export default function LanguagesSelector() {
	const { srcLangOptions, targetLangOptions, SRC_LANG, TRGT_LANG } = constants;

	const { dispatchSetLangCode, dispatchSetTrgtLangCode } = useTranslatorSlice();

	useEffect(() => {
		let mounted = true;

		if (mounted) dispatchSetLangCode('EN-GB');

		return () => (mounted = false);
	}, []);

	return (
		<Wrapper>
			<LanguageSelect langType={SRC_LANG} langOptions={srcLangOptions} />
			<SvgWrapper>
				<svg focusable="false" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
					<path d="M6.99 11L3 15l3.99 4v-3H14v-2H6.99v-3zM21 9l-3.99-4v3H10v2h7.01v3L21 9z"></path>
				</svg>
			</SvgWrapper>
			<LanguageSelect
				langType={TRGT_LANG}
				langOptions={targetLangOptions}
				onChange={(e) => dispatchSetTrgtLangCode(e.target.value)}
			/>
		</Wrapper>
	);
}

const Wrapper = styled.div`
	display: flex;
`;

const SvgWrapper = styled.div`
	height: 24px;
	width: 24px;
	fill: #70757a;
	display: inline-block;
	position: relative;
	margin-right: 20px;
	margin-top: 5px;
`;
