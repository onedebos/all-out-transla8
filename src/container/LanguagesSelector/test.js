/**
 * @jest-environment jsdom
 */
import { render, cleanup, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../utils/slices/store';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import LanguagesSelector from './';
import constants from './constants';

afterEach(() => {
	cleanup();
});

describe('<LanguagesSelector />', () => {
	const { SRC_LANG, TRGT_LANG } = constants;

	it('renders the LanguagesSelector component successfully without crashing', () => {
		render(
			<Provider store={store}>
				<LanguagesSelector />
			</Provider>
		);
	});

	it('allows the user select a different target language like German from the dropdown', () => {
		const { getByTestId, getAllByTestId } = render(
			<Provider store={store}>
				<LanguagesSelector />
			</Provider>
		);

		fireEvent.change(getByTestId(TRGT_LANG), { target: { value: 'DE' } });
		let options = getAllByTestId(TRGT_LANG + '-option');
		expect(options[0].selected).toBeFalsy();
		expect(options[1].selected).toBeFalsy();
		expect(options[2].selected).toBeFalsy();
		expect(options[3].selected).toBeTruthy();
	});

	it('allows the user select only English for the Source text', () => {
		const { getByTestId, getAllByTestId } = render(
			<Provider store={store}>
				<LanguagesSelector />
			</Provider>
		);
		let options = getAllByTestId(SRC_LANG + '-option');
		expect(options.length).toBe(1);
		expect(options[0]).toHaveValue('EN');
	});

	it('swaps the positions of the source and target dropdowns when the user clicks on the arrows', () => {
		const { getByTestId } = render(
			<Provider store={store}>
				<LanguagesSelector />
			</Provider>
		);
		const arrowBtn = getByTestId('arrow-btn');
		fireEvent.click(arrowBtn);

		const languagesWrapper = getByTestId('languages-wrapper');
		expect(languagesWrapper).toHaveStyle({
			display: 'flex',
			'flex-direction': 'row-reverse',
			'justify-content': 'flex-end',
		});
	});

	it('matches snapshot', () => {
		const tree = renderer
			.create(
				<Provider store={store}>
					<LanguagesSelector />
				</Provider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
