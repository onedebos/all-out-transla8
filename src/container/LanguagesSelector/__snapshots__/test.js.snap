// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguagesSelector /> matches snapshot 1`] = `
<div
  className="LanguagesSelector__Wrapper-sc-1ww3ozx-0 hvFoEs"
>
  <div
    className="LanguageSelect__SelectWrapper-sc-1umrqq7-0 iZfAxW"
  >
    <select
      aria-label="select-language"
      className="LanguageSelect__LanguageSelectStyle-sc-1umrqq7-1 nRspY"
      data-testid="SourceLanguage"
      name="SourceLanguage"
    >
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="SourceLanguage-option"
        value="EN"
      >
        English
      </option>
    </select>
  </div>
  <div
    className="LanguagesSelector__SvgWrapper-sc-1ww3ozx-1 FPuDG"
  >
    <svg
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.99 11L3 15l3.99 4v-3H14v-2H6.99v-3zM21 9l-3.99-4v3H10v2h7.01v3L21 9z"
      />
    </svg>
  </div>
  <div
    className="LanguageSelect__SelectWrapper-sc-1umrqq7-0 iZfAxW"
  >
    <select
      aria-label="select-language"
      className="LanguageSelect__LanguageSelectStyle-sc-1umrqq7-1 nRspY"
      data-testid="TargetLanguage"
      name="TargetLanguage"
      onChange={[Function]}
    >
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="BG"
      >
        Bulgarian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="CS"
      >
        Czech
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="DA"
      >
        Danish
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="DE"
      >
        German
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="EL"
      >
        Greek
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="EN-GB"
      >
        English (British)
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="EN-US"
      >
        English (American)
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="ES"
      >
        Spanish
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="ET"
      >
        Estonian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="FI"
      >
        Finnish
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="FR"
      >
        French
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="HU"
      >
        Hungarian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="IT"
      >
        Italian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="JA"
      >
        Japanese
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="LT"
      >
        Lithuanian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="LV"
      >
        Latvian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="NL"
      >
        Dutch
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="PL"
      >
        Polish
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="PT-PT"
      >
        Portuguese
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="PT-BR"
      >
        Portuguese (Brazilian)
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="RO"
      >
        Romanian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="RU"
      >
        Russian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="SK"
      >
        Slovak
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="SL"
      >
        Slovenian
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="SV"
      >
        Swedish
      </option>
      <option
        className="LanguageSelect__LangStyleOption-sc-1umrqq7-2"
        data-testid="TargetLanguage-option"
        value="ZH"
      >
        Chinese
      </option>
    </select>
  </div>
</div>
`;
